### PR TITLE
 ISPN-10161 Write custom surefire report for leaked threads

### PR DESCRIPTION
**Pull request by danberindei**
https://issues.jboss.org/browse/ISPN-10161

* Make JUnit report writer independent of TestNG
* Log hanged tests and leaked threads in test-failures log
  process_trace_logs.sh ignores test logs without a message
  in the test-failures log.
* Report the stack leaked threads and their creators better
* Write custom surefire report for leaked threads

### CODE REVIEW

**Comment by danberindei** [comment]
Leak error looks great: https://ci.infinispan.org/job/Infinispan/job/PR-7107/3/testReport/junit/org.infinispan.functional/FunctionalScatteredInMemoryTest/ThreadLeakChecker/

BTW, the leak is from https://issues.jboss.org/browse/ISPN-9988

File: commons-test/src/main/java/org/infinispan/commons/test/PolarionJUnitXMLReporter.java [hunk 1]
```diff
@@ -174,121 +156,79 @@ protected void generateReport(ISuite suite) {
...
-            generateFileName(suite) + ".xml", document.toXML());
+      PolarionJUnitXMLWriter writer = null;
+      try {
+         String outputDir = suite.getOutputDirectory().replaceAll(".Surefire suite", "");
+         String baseName = String.format("TEST-%s.xml", getSuiteName(suite));
+         File outputFile = new File(outputDir, baseName);
+         writer = new PolarionJUnitXMLWriter(outputFile);
+         writer.start(getModuleSuffix(), testCount, m_numSkipped.get(), m_numFailed.get(), elapsedTime, true);
+
+         // TODO Also write a report based on suite.getAllInvokedMethod() and check for differences
```
Comment (ryanemerson) [inline commons-test/src/main/java/org/infinispan/commons/test/PolarionJUnitXMLReporter.java]: Is this meant to be addressed in this PR?

File: commons-test/src/main/java/org/infinispan/commons/test/PrettyXMLStreamWriter.java [hunk 2]
```diff
@@ -0,0 +1,219 @@
+package org.infinispan.commons.test;
+
+import java.util.Arrays;
+import javax.xml.namespace.NamespaceContext;
+import javax.xml.stream.XMLStreamException;
+import javax.xml.stream.XMLStreamWriter;
+
+/**
+ *
+ */
```
Comment (ryanemerson) [inline commons-test/src/main/java/org/infinispan/commons/test/PrettyXMLStreamWriter.java]: Empty javadoc

File: core/src/test/java/org/infinispan/scattered/statetransfer/PrefetchTest.java [hunk 3]
```diff
@@ -68,6 +69,7 @@ public void testPrefetch12() throws InterruptedException, BrokenBarrierException
    }
 
    private void testPrefetch(int invokePhase, int receivePhase) throws InterruptedException, TimeoutException, BrokenBarrierException, ExecutionException {
+      new Thread(() -> TestingUtil.sleepThread(100000), "name").start();
```
Comment (ryanemerson) [inline core/src/test/java/org/infinispan/scattered/statetransfer/PrefetchTest.java]: Should this be a separate commit?

File: commons-test/src/main/java/org/infinispan/commons/test/PolarionJUnitXMLReporter.java (hunk 1 above)
Comment (danberindei) [reply commons-test/src/main/java/org/infinispan/commons/test/PolarionJUnitXMLReporter.java]: Nope, it just isn't clear why we need to keep track of the executed tests ourselves instead of using TestNG's built-in results

File: core/src/test/java/org/infinispan/scattered/statetransfer/PrefetchTest.java (hunk 3 above)
Comment (danberindei) [reply core/src/test/java/org/infinispan/scattered/statetransfer/PrefetchTest.java]: It shouldn't be in the PR at all :)

File: commons-test/src/main/java/org/infinispan/commons/test/PrettyXMLStreamWriter.java (hunk 2 above)
Comment (danberindei) [reply commons-test/src/main/java/org/infinispan/commons/test/PrettyXMLStreamWriter.java]: Comment added

**Comment by danberindei** [comment]
Updated @ryanemerson

**Comment by danberindei** [comment]
Rebased and ignored the new thread leaks in AuthenticationLDAPIT

**Comment by karesti** [comment]
@danberindei is this ok to be merged ?

**Comment by ryanemerson** [comment]
@danberindei What's the status of this PR?

**Comment by danberindei** [comment]
Should be ok to merge @ryanemerson @karesti, I'll push again to trigger a new build just in case

**Comment by ryanemerson** [comment]
@danberindei I'm not sure if this is expected or not with these changes:

```

ThreadLeakCheckerorg.infinispan.commons.test.ThreadLeakChecker$LeakException: Leaked thread: pool-5-thread-1 << UNKNOWN[null] at java.base@11.0.3/jdk.internal.misc.Unsafe.park(Native Method) at java.base@11.0.3/java.util.concurrent.locks.LockSupport.parkNanos(LockSupport.java:234) at java.base@11.0.3/java.util.concurrent.locks.AbstractQueuedSynchronizer$ConditionObject.awaitNanos(AbstractQueuedSynchronizer.java:2123) at java.base@11.0.3/java.util.concurrent.ScheduledThreadPoolExecutor$DelayedWorkQueue.take(ScheduledThreadPoolExecutor.java:1182) at java.base@11.0.3/java.util.concurrent.ScheduledThreadPoolExecutor$DelayedWorkQueue.take(ScheduledThreadPoolExecutor.java:899) at java.base@11.0.3/java.util.concurrent.ThreadPoolExecutor.getTask(ThreadPoolExecutor.java:1054) at java.base@11.0.3/java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1114) at java.base@11.0.3/java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:628) at java.base@11.0.3/java.lang.Thread.run(Thread.java:834) Caused by: org.infinispan.commons.test.ThreadLeakChecker$LeakException: UNKNOWN[null]
--



```

**Comment by danberindei** [comment]
@ryanemerson I had to add an exclusion in `LdapServerRule` because ApacheDS leaks a thread, but I think this one is because some server-runtime tests failed to start (and Jenkins ignored them).

**Comment by tristantarrant** [comment]
Merged
